Structure_Engine/Modify/Flip.cs

```diff
@@ -47,15 +58,89 @@ public static Bar Flip(this Bar bar)
 
             Bar flipped = bar.ShallowClone();
 
+            // Flip Nodes
             Node tempNode = flipped.Start;
             flipped.Start = flipped.End;
             flipped.End = tempNode;
 
+            if (bar.IsVertical())
+                flipped.OrientationAngle = -bar.OrientationAngle + Math.PI;
+            else
+                flipped.OrientationAngle = -bar.OrientationAngle;
+
+            // Flip releases
+            BarRelease flippedRelease = bar.Release?.ShallowClone();
+            if (flippedRelease != null)
+            {
+                Constraint6DOF tempRelease = flippedRelease.StartRelease;
+                flippedRelease.StartRelease = flippedRelease.EndRelease;
+                flippedRelease.EndRelease = tempRelease;
+                flipped.Release = flippedRelease;
+            }
+
+            // Flip section property
+            if (bar.SectionProperty != null)
+            {
+                ISectionProperty flippedSectionProperty = Flip(bar.SectionProperty);
+                flipped.SectionProperty = flippedSectionProperty;
+            }
+
+            // Flip offsets
+            if (bar.Offset != null)
+            {
+                Offset flippedOffset = new Offset()
+                {
+                    Start = new Vector()
+                    {
+                        X = -bar.Offset.End.X,
+                        Y = -bar.Offset.End.Y,
+                        Z = bar.Offset.End.Z
+                    },
+                    End = new Vector()
+                    {
+                        X = -bar.Offset.Start.X,
+                        Y = -bar.Offset.Start.Y,
+                        Z = bar.Offset.Start.Z
+                    }
+                };
+
+                flipped.Offset = flippedOffset;
+            }
+
+
+
             return flipped;
         }
 
         /***************************************************/
 
+        [Description("Flips the curve layout, i.e. the start becomes the end and vice versa.")]
+        [Input("layout", "The curvlayout to flip.")]
+        [Output("layout", "The layout with a flipped location curve.")]
+        public static ICurveLayout Flip(this ICurveLayout layout)
+        {
+            if (layout.IsNull())
+                return null;
+
+            if (layout is ExplicitCurveLayout)
+            {
+                ExplicitCurveLayout explicitLayout = layout as ExplicitCurveLayout;
+                return new ExplicitCurveLayout(explicitLayout.Curves.Select(x => x.IMirror(Plane.YZ)));
+
+            }
+            else if (layout is OffsetCurveLayout)
+            {
+                return layout;
+            }
+            else
+            {
+                Engine.Base.Compute.RecordError("ICurveLayout type not recognised.");
+                return null;
```

Comment:
Would jsut return the unflipped verion with the error. Better than null I think

```suggestion
                Engine.Base.Compute.RecordError("ICurveLayout type not recognised. Unflipped curve layout returned");
                return layout;
```

Reply:
Agreed - will close.